[Gear] Remade shadowgrasp trinket into a pet spawn.

adrian-budau opened the pull request:
Also slightly changed the way it targets enemies since it seems to
auto-update based on player target.

adrian-budau committed:
[Gear] Remade shadowgrasp trinket into a pet spawn.

Also slightly changed the way it targets enemies since it seems to
auto-update based on player target.

EvanMichaels commented:
Looks nice, this was on the to-do for a while. Do you happen to have some sim comparisons to confirm the numbers look good with the changes compared to previously?

adrian-budau commented:
Yeah, windwalkers have Hit Combo which increases pet damage (by up to 6%).

This is before:
![image](https://user-images.githubusercontent.com/1350273/122621367-fbed2e00-d09d-11eb-9246-d421cc513f9d.png)

And this is after:
![image](https://user-images.githubusercontent.com/1350273/122621382-0ad3e080-d09e-11eb-9f1c-57704496dd95.png)

The increase is almost 6% (First few hits of the sim only get 2-5% of pet damage increase because hit combo isn't stacked yet).

EvanMichaels commented:
Cool, just wanted to confirm the executes were the same, that the haste increases were working, etc.

I assume changing it to a pet also implicitly (and more cleanly) fixed the issue I had the workaround for with the target multipliers?

adrian-budau commented:
Yeah, it benefits from pet target multipliers instead of player target multiplier (hence why I removed that logic).

adrian-budau commented:
Is there anything more I need to add to this PR?

gastank commented on engine/player/unique_gear_shadowlands.cpp:
```diff
@@ -1681,45 +1691,50 @@ void shadowgrasp_totem( special_effect_t& effect )
+    }
+  };
+
+  effect.execute_action = create_proc_action<shadowgrasp_totem_t>( "shadowgrasp_totem", effect );
```
You don't need to create an action just to trigger the buff. You can retain the old code with the sanity check for buff already existing and instead override `shadowgrasp_totem_buff_t::trigger()` to spawn the pet then continue with `buff_t::trigger()`

adrian-budau replied:
Yes, sorry. Initially I wanted to make sure the pet was spawning before the buff ticks (in case they ever make the buff tick on spawn, now it doesn't), but doing it like you said works as well, I'm just not familiar with all the simc hierarchy.

adrian-budau committed:
Fixed a typo, and removed the extra action in implementation.

adrian-budau commented:
Removed the extra action, fixed a typo and actually put  the npc id
![image](https://user-images.githubusercontent.com/1350273/123666075-91788280-d841-11eb-82d5-dc5007e761ac.png)

gastank commented:
> 
> 
> Should I move it outside?

Yeah it was likely an oversight with the initial implementation. Sorry took a while but we were all busy fighting over gromit spawns. If you can move it out we'll get the PR merged.

adrian-budau committed:
Review fix

adrian-budau commented:
Sorry I didn't notice the reply (too many rares & mounts in Korthia&Maw). Fixed it :-D
